### PR TITLE
Extend run_jedi_exe.py to run various fv3-jedi apps

### LINKED ISSUE #223: Extend functionality of run_jedi_exe.py
Make run_jedi_exe script more generic/flexible to exercise more components

### PR DESCRIPTION
**Pull request by RussTreadon-NOAA**
Issue #223 documents updates which increase the types of fv3-jedi executables `run_jedi_exe.py` can launch.   The updated script can launch the following fv3-jedi exeucutables in the following modes:
- `fv3jedi_hofx_nomodel.x`  for 3dhofx  
- `fv3jedi_hofx.x`  for 4dhofx 
- `fv3jedi_var.x`  for 3dvar
- `fv3jedi_letkf.x` for lgetkf

Fixes #223

### CODE REVIEW

**Comment by guillaumevernieres** [comment]
@RussTreadon-NOAA , is there a reason to run the `RT` and `GW-RT` tests? I thought all the ctests were included in `GW-RT`?

**Comment by RussTreadon-NOAA** [comment]
> @RussTreadon-NOAA , is there a reason to run the `RT` and `GW-RT` tests? I thought all the ctests were included in `GW-RT`?

You are right @guillaumevernieres .   I got label happy & wanted to see more green `PASSED` labels.  No reason to repeat `RT` when it's a subset of `GW-RT`.

File: parm/atm/obs/config/lgetkf_amsua_n19.yaml [hunk 1]
```diff
@@ -26,6 +26,36 @@ obs operator:
     CoefficientPath: ./crtm/
 obs error:
   covariance model: diagonal
+obs bias:
+  input file: !ENV ${DATA}/obs/${GPREFIX}amsua_n19.satbias.${GDATE}.nc4
+  output file: !ENV ${DATA}/bc/${APREFIX}amsua_n19.satbias.${CDATE}.nc4
```
Comment (AndrewEichmann-NOAA) [inline parm/atm/obs/config/lgetkf_amsua_n19.yaml]: Use of `$CDATE` within global-workflow is deprecated (except when locally assigned) in favor of `${PDY}${cyc}`. I think the same applies to `$GDATE`

**Comment by RussTreadon-NOAA** [comment]
Investigation into the `test_gdasapp_run_jedi_exe_letkf ` failure shows that `ufsda.r2d2.fetch` is unable to find the amsua_n19 satbias_cov file.
```
fetch
  date:
           20211221000000
  default_user_date_format: %Y-%m-%dT%H:%M:%SZ
  experiment: oper_gdas
  file_format:
  file_type:
            satbias_cov
  ignore_missing: False
  obs_type:
           amsua_n19
  provider: gsi
  target_file: ./obs/gdas.t00z.amsua_n19.satbias_cov.2021122100.nc4
  type:     bc
  user_date_format:

2023-02-04 00:14:03,782 - INFO - r2d2 - Starting
2023-02-04 00:14:03,784 - INFO - localdb - local: fetch from /scratch2/NCEPDEV/stmp1/Cory.R.Martin/R2D2/gsi/bc/oper_gdas/2021-12-21/gsi.oper_gdas.bc.amsua_n19.2021-12-21T00:00:00Z.satbias_cov to ./obs/gdas.t00z.amsua_n19.satbias_cov.2021122100.nc4
2023-02-04 00:14:03,784 - INFO - localdb - shared: fetch from /scratch1/NCEPDEV/stmp4/Cory.R.Martin/R2D2_SHARED/gsi/bc/oper_gdas/2021-12-21/gsi.oper_gdas.bc.amsua_n19.2021-12-21T00:00:00Z.satbias_cov to ./obs/gdas.t00z.amsua_n19.satbias_cov.2021122100.nc4
FileNotFoundError: Could not find gsi.oper_gdas.bc.amsua_n19.2021-12-21T00:00:00Z.satbias_cov in local or shared
+ rc=1
```

This is a valid failure.    The amsua_n19 satbias_cov file is in neither of the searched R2D2 paths.   

The error did not occur in my command line execution of `test_gdasapp_run_jedi_exe_letkf` because my R2D2 database has a valid amsua_n18 satbias_cov file. 
```
fetch
  date:
           20211221000000
  default_user_date_format: %Y-%m-%dT%H:%M:%SZ
  experiment: oper_gdas
  file_format:
  file_type:
            satbias_cov
  ignore_missing: False
  obs_type:
           amsua_n19
  provider: gsi
  target_file: ./obs/gdas.t00z.amsua_n19.satbias_cov.2021122100.nc4
  type:     bc
  user_date_format:

2023-02-04 11:30:50,284 - INFO - r2d2 - Starting
2023-02-04 11:30:50,285 - INFO - localdb - local: fetch from /scratch2/NCEPDEV/stmp1/Russ.Treadon/R2D2/gsi/bc/oper_gdas/2021-12-21/gsi.oper_gdas.bc.amsua_n19.2021-12-21T00:00:00Z.satbias_cov to ./obs/gdas.t00z.amsua_n19.satbias_cov.2021122100.nc4
2023-02-04 11:30:50,287 - INFO - r2d2 - Success
```
Examination of the satbias_cov section of `ush/ufsda/stage.py` found the following hack in function `gdas_single_cycle`
```
            r2d2_config['file_type'] = 'satbias_cov'
            target_file2 = target_file.replace('satbias', 'satbias_cov')
            r2d2_config['target_file_fmt'] = target_file2
            try:
                ufsda.r2d2.fetch(r2d2_config)
            except FileNotFoundError:
                logging.error("Warning: satbias_cov file cannot be fetched from R2D2!")
                # temp hack to copy satbias as satbias_cov
                # if satbias_cov does not exists in R2D2
                if os.path.isfile(target_file) and not os.path.isfile(target_file2):
                    shutil.copy(target_file, target_file2)
```
A check of git history traces the above addition to GDASApp PR #63.    This hack is not in function `bias_obs` of `stage.py`   The letkf execution of `run_jedi_exe.py` uses `bias_obs` to populate the run directory with bias correction files.

We should remove the satibas_cov hack in `stage.py`.   This may naturally come as we move away from using R2D2 for atmospheric UFS DA.    Since this remains future work it is beneficial, at present, to populate a shared R2D2 database with the expected bias correction files (satbias.nc4, satbias_cov.nc4, satbias_tlapse.txt).

Taking a step back, does the letkf application need satbias_cov?   A similar question applies to hofx.   I think not.   The satbias_cov file is used for variational bias correction in var applications.   

Thus, another way to address the above failure is to remove satbias_cov from the amsua_n19 observation yaml used by letkf.   This is easily done but it means that var and letkf/hofx applications can not use the same observation yaml.  Maintaining two nearly but not totally identical yamls is problematic.  Updates to one may be missed in the other.

For the time being the less than ideal patch of adding the satbias_cov hack to function `bias_obs` in `stage.py` was taken.   Done at 70b7d65.

**Comment by RussTreadon-NOAA** [comment]
Orion `test_gdasapp_run_jedi_exe_letkf` failed due to long running job.  Suspect disk issue.  Rerun GW-RT on Orion.

**Comment by RussTreadon-NOAA** [comment]
@CoryMartin-NOAA , Orion GW-RT tests keep failing to finish.   A check of `/work2/noaa/stmp/cmartin/CI/GDASApp/workflow/PR/311/global-workflow/sorc/gdas.cd` indicates that `build.sh` did not run.   Does the Orion GW-RT test generate a log file?  If so, where is it?  I'd like to check it to see what's happening or not happening.

**Comment by CoryMartin-NOAA** [comment]
@RussTreadon-NOAA there is supposed to be a build log for the workflow build_all.sh command but it looks like it isn't even getting that far. 

I suspect though, that the issue here is a flaw in my CI system design. I don't think it is doing anything because there is not an updated commit hash. That's the first thing the script checks, since there is not a newer commit, it just exits the cron early. Perhaps there needs to be some sort of lock file that it checks for that says 'passed' or 'failed' and will rerun if 'failed'? In the meantime, which would you prefer? That I delete the directory and reassign the labels? Or push an empty commit to this repo and reassign the labels?

**Comment by RussTreadon-NOAA** [comment]
@CoryMartin-NOAA , thank you for explaining what's going on.   Given your reply I now see
```
  # get commit hash
  commit=$(git log --pretty=format:'%h' -n 1)
  if [ -f "$GDAS_CI_ROOT/PR/$pr/commit" ]; then
    oldcommit=$(cat $GDAS_CI_ROOT/PR/$pr/commit)
    if [ $oldcommit == $commit ]; then
      # do no more for this PR, as the commit has already been tested
      continue
    fi
  fi
  echo "$commit" > $GDAS_CI_ROOT/PR/$pr/commit
```
in `ci/driver.sh` and `gw-driver.sh`

If we 

1. remove `orion-GW-RT-Running` label from this PR
2. `rm commit` from `/work2/noaa/stmp/cmartin/CI/GDASApp/workflow/PR/311`
3. add `orion-GW-RT` to this PR

would this suffice to fully rerun Orion GW-RT for this PR?

Here's a thought.

The GW-RT label must be manually added by the develop to trigger the GW-RT tests, right?   If so, what do you think about assuming that when a developer applies the label (s)he intends to run or rerun the GW-RT test?   This means we do not check the commit hash when executing GW-RT.

**Comment by CoryMartin-NOAA** [comment]
@RussTreadon-NOAA yes I think that makes sense, I think this was a case of me trying to cover too many bases when the simpler solution would've worked well enough. Let me `rm` right now and then reapply the label. We can fix this in the CI script in an upcoming PR.

**Comment by RussTreadon-NOAA** [comment]
Thanks @CoryMartin-NOAA for removing `commit` and requeuing Orion GW-RT.   

I'm beginning to feel that I over engineered the current `run_jedi_exe.py` ctests.   At present
```
  Test #1341: test_gdasapp_run_jedi_exe_3dhofx
  Test #1342: test_gdasapp_run_jedi_exe_3dvar
  Test #1343: test_gdasapp_run_jedi_exe_letkf
```
not only execute `run_jedi_exe.py` for the indicated mode.  They also monitor the submitted job and wait for it to successfully run to completion before returning a `Passed`.   Each test takes 200 to 300 seconds to run.    Other ctests run in much less time.   

The `test_gdasapp_atm_jjob` suite of ctests execute `fv3jedi_var.x` and `fv3jedi_letkf.x`.    Both executions involve hofx calculations.   Given this, I could and probably should simplify the `run_jedi_exe.py` ctests to include three checks and one action
1. ensure `run_jedi_exe.py` runs without error
2. confirm that the generated yaml is a valid yaml file
3. check for valid job submission (no sbatch errors)
4. scancel the submitted job

Sound reasonable?

**Comment by CoryMartin-NOAA** [comment]
@RussTreadon-NOAA yes the current testing is probably overkill.  If these unit tests complete successfully, then it is reasonable to assume that the MPI job will execute successfully, especially if there are other tests for that.

**Comment by RussTreadon-NOAA** [comment]
9dbc973 reorganizes the `run_jedi_exe` ctests as outlined in the comment [above](https://github.com/NOAA-EMC/GDASApp/pull/311#issuecomment-1419426218):
- move `run_jedi_exe` ctests up one directory from `atm/global-workflow` to `atm` since these tests do not depend on g-w
- simply `run_jedi_exe` ctests to (1) execute `run_jedi_exe.py` for specified mode, (2) check for valid job submission and then cancel the job, and (3) confirm generation of valid yaml files.

Revised `run_jedi_exe` ctests run much faster on Hera
```
(gdasapp) Hera(hfe03):/scratch1/NCEPDEV/da/Russ.Treadon/git/global-workflow/work/sorc/gdas.cd/build$ ctest -R test_gdasapp_run_jedi_exe
Test project /scratch1/NCEPDEV/da/Russ.Treadon/git/global-workflow/work/sorc/gdas.cd/build
    Start 1327: test_gdasapp_run_jedi_exe_3dhofx
1/3 Test #1327: test_gdasapp_run_jedi_exe_3dhofx ...   Passed   19.50 sec
    Start 1328: test_gdasapp_run_jedi_exe_3dvar
2/3 Test #1328: test_gdasapp_run_jedi_exe_3dvar ....   Passed   10.07 sec
    Start 1329: test_gdasapp_run_jedi_exe_letkf
3/3 Test #1329: test_gdasapp_run_jedi_exe_letkf ....   Passed    5.97 sec

100% tests passed, 0 tests failed out of 3

Total Test time (real) =  35.83 sec
```
and Orion
```
(gdasapp) Orion-login-1:/work/noaa/da/rtreadon/git/global_workflow/work/sorc/gdas.cd/build$ ctest -R test_gdasapp_run_jedi_exe
Test project /work/noaa/da/rtreadon/git/global_workflow/work/sorc/gdas.cd/build
    Start 1327: test_gdasapp_run_jedi_exe_3dhofx
1/3 Test #1327: test_gdasapp_run_jedi_exe_3dhofx ........   Passed   82.05 sec
      Start 1328: test_gdasapp_run_jedi_exe_3dvar
2/3 Test #1328: test_gdasapp_run_jedi_exe_3dvar .........   Passed    8.16 sec
      Start 1329: test_gdasapp_run_jedi_exe_letkf
3/3 Test #1329: test_gdasapp_run_jedi_exe_letkf .........   Passed    6.10 sec
```

**Comment by RussTreadon-NOAA** [comment]
Hera and Orion GW-RT tests finally pass!
   
@AndrewEichmann-NOAA , I do not anticipate making any more commits to [`feature/runexe`](https://github.com/NOAA-EMC/GDASApp/tree/feature/runexe) unless your review requests changes.

**Comment by AndrewEichmann-NOAA** [comment]
@RussTreadon-NOAA I won't block on the CDATE thing, but others might later.

**Review by AndrewEichmann-NOAA — DISMISSED** [review]

File: parm/atm/obs/config/lgetkf_amsua_n19.yaml (hunk 1 above)
Comment (RussTreadon-NOAA) [reply parm/atm/obs/config/lgetkf_amsua_n19.yaml]: Valid point.  `${CDATE}` and `${GDATE}` have been replaced with `${PDY}${cyc}` and `${gPDY}${gcyc}`, respectively in files touched by this PR.  

Changes to `ush/ufsda/genYAML.py`  required replacement of `CDATE` in `test/genYAML_prep.sh` with `PDY` and `cyc`   Updates were also required in `parm/atm/obs/config` files `amsua_n19.yaml`, `lgetkf_amsua_n19.yaml`, `lgetkf_sondes.yaml`, and `sondes.yaml` as well as `ush/ufsda/misc_utils.py`

Done at b349d32.


Additional `ush/` and `parm/` files remain with references to `${CDATE}` and `${GDATE}`.   Replacing these can be the subject of a new issue and PR.

**Comment by RussTreadon-NOAA** [comment]
Thanks again, @AndrewEichmann-NOAA , for the useful feedback.   I'm glad to hear that `CDATE` and `GDATE` are begin removed from g-w.  Operations uses `PDY` and `cyc`.  We should too.

As noted above [b349d32](https://github.com/NOAA-EMC/GDASApp/pull/311/commits/b349d3235a2d615a3cfd7cb18e49e853eed1042c) does not remove all occurrences of `CDATE` and `GDATE` in GDASApp.  This work along with some other items mentioned in this PR can be the the subject of new GDASApp issues.

If you are satisfied with the changes at [b349d32](https://github.com/NOAA-EMC/GDASApp/pull/311/commits/b349d3235a2d615a3cfd7cb18e49e853eed1042c), I'll resolve the conversation.

**Comment by RussTreadon-NOAA** [comment]
GDASApp issue #330 created to address `${CDATE}` and `${GDATE}` usage in [NOAA-EMC/GDASApp](https://github.com/NOAA-EMC/GDASApp).

**Comment by RussTreadon-NOAA** [comment]
> @RussTreadon-NOAA yes I think that makes sense, I think this was a case of me trying to cover too many bases when the simpler solution would've worked well enough. Let me `rm` right now and then reapply the label. We can fix this in the CI script in an upcoming PR.

@CoryMartin-NOAA , which CI script should we update:  `ci/driver.sh` or `ci/gw_driver.sh`?  Both contain 
```
    if [ $oldcommit == $commit ]; then
      # do no more for this PR, as the commit has already been tested
      continue
    fi
```
logic

**Comment by CoryMartin-NOAA** [comment]
@RussTreadon-NOAA I guess both? the former is for the labels like `hera-RT` and the latter is for `hera-GW-RT`

**Comment by RussTreadon-NOAA** [comment]
> @RussTreadon-NOAA I guess both? the former is for the labels like `hera-RT` and the latter is for `hera-GW-RT`

OK.  Let me remove the `oldcommit` check from both, leaving

`driver.sh`
```
  # get commit hash
  commit=$(git log --pretty=format:'%h' -n 1)
  echo "$commit" > $GDAS_CI_ROOT/PR/$pr/commit
```
`gw_driver.sh`
```
  # get commit hash
  commit=$(git log --pretty=format:'%h' -n 1)
  echo "$commit" > $GDAS_CI_ROOT/workflow/PR/$pr/commit
```

**Review by AndrewEichmann-NOAA — APPROVED** [review]

**Comment by RussTreadon-NOAA** [comment]
GW-RT tests on Hera and Orion pass for `feature/run_exe` at [`39126ae`](https://github.com/NOAA-EMC/GDASApp/pull/311/commits/39126ae06bcc0b31bfe4ad62ec4ebcf6a58a265c).    Given this plus @AndrewEichmann-NOAA's  `approve`, I'll merge this PR into `develop`.

**Pull request merged**